### PR TITLE
Re-initialize diag functors after LoadBalance

### PR DESCRIPTION
**Pull request by RevathiJambunathan**
This PR addresses the issue #940 , where a segfault occured in diagnostics after load balancing, due to dangling pointers.

To fix this issue, In this PR, the diagnostic functors are initialized within a function -- "InitializeFieldFunctors()".
When load balancing is performed and if a new partitioning is generated, then this new function "InitializeFieldFunctors()" is called from Regrid, if (doRedistribute==true). 
 This ensures that the source MultiFabs used in the new diagnostics points to the correct field MultiFab since they were re-allocated in Regrid.cpp.

Tests:
Performed the reduced_diags test again with load balancing using the input file and parameters given in Issue#940 and compared old and new plotfiles:

./fcompare.gnu.ex diags/plotfiles/plt00003/ diags/d00003/

confirmed that the absolute and relative error is 0 for the fields and current.

### CODE REVIEW

File: Source/Diagnostics/Diagnostics.cpp [hunk 1]
```diff
@@ -376,3 +330,59 @@ Diagnostics::DefineDiagMultiFab ( int lev ) {
...
+                m_all_field_functors[lev][comp] = new PartPerGridFunctor(nullptr, lev, m_crse_ratio);
+            } else if ( m_varnames[comp] == "divB" ){
+                m_all_field_functors[lev][comp] = new DivBFunctor(warpx.get_array_Bfield_aux(lev), lev, m_crse_ratio);
+            } else if ( m_varnames[comp] == "divE" ){
+                m_all_field_functors[lev][comp] = new DivEFunctor(warpx.get_array_Efield_aux(lev), lev, m_crse_ratio);
+            }
+        }
+
+        AddRZModesToDiags( lev );
+        // At this point, m_varnames.size() >= m_all_field_functors[0].size()
```
Comment (MaxThevenet) [inline Source/Diagnostics/Diagnostics.cpp:386]:
```suggestion
```

File: Source/Parallelization/WarpXRegrid.cpp [hunk 2]
```diff
@@ -70,6 +70,8 @@ WarpX::LoadBalance ()
     }
     if (doRedistribute)
     {
+        // Re-initialize diagnostic functors that stores pointers to the user-requested fields.
+        multi_diags->InitializeFieldFunctors();
```
Comment (ax3l) [inline Source/Parallelization/WarpXRegrid.cpp:74]: are the old functors destroyed at this point? Just checking, because they do hold quite an amount of dynamically allocated data members with `new` and I cannot quickly spot the `delete` counter parts.

File: Source/Diagnostics/Diagnostics.cpp [hunk 3]
```diff
@@ -89,61 +89,15 @@ Diagnostics::InitData ()
...
-
-        // At this point, m_varnames.size() >= m_all_field_functors[0].size()
-
         // Initialize member variable m_mf_output depending on m_crse_ratio, m_lo and m_hi
         DefineDiagMultiFab( lev );
-
     }
+
+    InitializeFieldFunctors ();
+
```
Comment (MaxThevenet) [inline Source/Diagnostics/Diagnostics.cpp:100]:
```suggestion

    // At this point, m_varnames.size() >= m_all_field_functors[0].size()

```

File: Source/Diagnostics/MultiDiagnostics.H [hunk 4]
```diff
@@ -19,6 +19,9 @@ public:
     void InitData ();
     /** \brief Called at each iteration. Compute diags and flush. */
     void FilterComputePackFlush (int step, bool force_flush=false);
+    /** \brief Loop over diags in all diags and call their InitializeFieldFunctors.
+               Called when a new partitioning is generated after load-balancing. */
```
Comment (MaxThevenet) [inline Source/Diagnostics/MultiDiagnostics.H:23]:
```suggestion
    /** \brief Loop over diags in all diags and call their InitializeFieldFunctors.
     * Called at initialization (to create the diagnostics functors) or when a new partitioning
     * is generated after load-balancing, because pointers to WarpX multifabs are dangling then.
     */
```

File: Source/Parallelization/WarpXRegrid.cpp (hunk 2 above)
Comment (MaxThevenet) [reply Source/Parallelization/WarpXRegrid.cpp:74]: Good point.

Comment (mrowan137) [reply Source/Parallelization/WarpXRegrid.cpp:74]: maybe `std::unique_ptr` is an option there.

File: Source/Diagnostics/MultiDiagnostics.H (hunk 4 above)
Comment (RevathiJambunathan) [reply Source/Diagnostics/MultiDiagnostics.H:23]: We do not call this function during initialization from MultiDiagnostics.H. We do, however, call InitializeFieldFunctors () in Diagnostics.H from InitData(), which is called from multi_diags. Do you think it would be accurate to mention that this function is called at initialization -- because as of now the loop of all diags to initializefieldfunctors is performed only after LB.

File: Source/Parallelization/WarpXRegrid.cpp (hunk 2 above)
Comment (RevathiJambunathan) [reply Source/Parallelization/WarpXRegrid.cpp:74]: Thank you @ax3l for the great point and for sharing your knowledge on this offline. As per our discussion, I have included some changes in the PR. @mrowan137 that is right! I have now included std::unique_ptr  in this PR. 
Thanks all!

File: Source/Diagnostics/MultiDiagnostics.H (hunk 4 above)
Comment (MaxThevenet) [reply Source/Diagnostics/MultiDiagnostics.H:23]: Oops no you're right my comment was inaccurate.

**Comment by RevathiJambunathan** [comment]
Thank you @MaxThevenet for the suggestions. Included them as commits -- except for one comment on function call at initialization. 

Regarding the call to InitializeFieldFunctors() from WarpXRegrid.cpp
In InitializeFieldFunctors() -- we resize the vector `m_all_field_functors` with nlev -- then loop over all levels to initialize the unique_ptrs to functors. 
When we call this function from RemakeLevel(int lev) -- we will be re-initializing functors for all levels everytime a level is being Remade.
 In other words, if RemakeLevel(lev) is called for lev=0, and lev=1, then the field_functors will be initialized for all levels both times. This repetition can be avoided.

Instead of calling Initialize for every level (potentially), we take advantage of the boolean `doRedistribute` that is toggled to true when RemakeLevel(lev) is performed. 
InitializeFieldFunctors is called if doRedistribute==true.

**Comment by RevathiJambunathan** [comment]
Update : After an online discussion with @MaxThevenet , it is safer to call the new InitializeFieldFunctors(int lev) from RemakeLevel(int lev) -- to ensure safe re-initialization -- in case the way we toggle `doRedistribute` changes in the future. 
Along the same lines, for simulations where the number of levels can dynamically change, we have decided to initialize the vector of diags with max_level -- which is a compulsory input parameter - -instead of finestLevel() + 1. 
These changes are [WIP]. Thank you @MaxThevenet for the suggestions!!!
